### PR TITLE
Fixes secret tag leaks

### LINKED ISSUE #2778: Tags cannot be secret
### What happened?

When marking tags as secret they actually aren't secret in the state:

```typescript
const bucket = new aws.s3.Bucket("my-bucket",  {tags: {"that": pulumi.secret("foo")}});
```

After

```
pulumi up --yes
pulumi stack export
``` 

Observe secret value in plaintext got copied to tags_all.

This happens because the TF integration layer copies to tags_all but does not understand secrets.

### Expected Behavior

Expected the secret to not be in the statefile. 

The behavior is a bit pedantic since in real world use cases tags are fairly exposed in the cloud, can't think of an easy use case where they should be secret, but it is certainly an odd corner case.

### Steps to reproduce

See above.

### Output of `pulumi about`

```
CLI          
Version      3.74.0
Go Version   go1.20.5
Go Compiler  gc

Plugins
NAME    VERSION
aws     6.0.3
aws     5.42.0
awsx    1.0.5
docker  3.6.1
nodejs  unknown

Host     
OS       darwin
Version  13.4.1
Arch     x86_64

This project is written in nodejs: executable='/Users/t0yv0/.nix-profile/bin/node' version='v18.16.1'

Current Stack: t0yv0/repro-aws-tagpanic/dev

Found no resources associated with dev

Found no pending operations associated with dev

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/t0yv0
User           t0yv0
Organizations  t0yv0, pulumi

Dependencies:
NAME            VERSION
@pulumi/aws     6.0.3
@pulumi/awsx    1.0.5
@pulumi/pulumi  3.80.0
@types/node     16.18.46

Pulumi locates its logs in /var/folders/gk/cchgxh512m72f_dmkcc3d09h0000gp/T/ by default
```

### Additional context

N/A

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by t0yv0**
Fixes #2778

Secrets no longer leak through tagsAll because tagsAll is removed before persisting to Pulumi state files.

### CODE REVIEW

File: examples/bucket-secret-tags-yaml/Pulumi.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+name: test-aws-2778-bucket-with-secret-tugs
+runtime: yaml
+resources:
```
[the remark this answers is not in the document]
Comment (t0yv0) [reply examples/bucket-secret-tags-yaml/Pulumi.yaml:3]: Can you elaborate? What do you mean?

[the remark this answers is not in the document]
Comment (iwahbe) [reply examples/bucket-secret-tags-yaml/Pulumi.yaml:3]: If the tag is a secret input to an explicit provider, it also needs to be secret in the state of resources that are provisioned with that provider.

**Comment by t0yv0** [comment]
Any runbook on license/snyk (Pulumi Open Source) failed check?

**Comment by iwahbe** [comment]
It's the same failure as https://github.com/pulumi/pulumi-aws/pull/2792. It's failing on GitHub.com/hashicorp/hcl, which is MPL 2.0. We should be ok, but we have decided to wait on @lukehoban's ok before merging.

**Review by iwahbe — APPROVED** [review]

**Comment by t0yv0** [comment]
Failure looked unrelated, I've rebased and will try again here.

**Comment by t0yv0** [comment]
I'll wait for https://github.com/pulumi/pulumi-aws/pull/2792 to go in and then rebase.

**Comment by t0yv0** [comment]
Hm this last round has some legitimate test failures on the newly introduced test as well as on some other tests. So this needs rework.

**Comment by t0yv0** [comment]
The test suite is pretty good.  This approach of removing tagsAll does not work.  The tests remind me that that'd be a breaking change to any user code that tries to read data out of tagsAll which may be reason enough not to do it, however there's more.

At first I thought the issue is that the tests are written with asserts on tagsAll which we could rewrite the tests to make asserts on tags. But that fails ALSO. There is a semantic failure where Pulumi fails to cleanly transition between these two states if tagsAll is removed:

```
examples/bucket-default-tags-yaml/step4 -> 
examples/bucket-default-tags-yaml/step5
```

So we can't remove tagsAll not just because we don't want to break programs that read their data but also because something in the diff method semantically relies on tagsAll. So tagsAll has to sit in Pulumi state, seemingly. Or else we need another hook or reuse one of the existing hooks to repopulate it to the copy of tags when reading Pulumi state.

**Comment by t0yv0** [comment]
IN addition to that, this program has a separate independent root cause of failure to not leak secrets:

```
name: test-aws-2778-bucket-with-secret-tugs
runtime: yaml
resources:

  aws-provider:
    type: pulumi:providers:aws
    options:
      defaultTags: 
        tags: 
          foo:
            fn::secret: mysecret            
  
  res:
    type: aws:s3:BucketV2
    options:
      provider: ${aws-provider}
```

This one has to do with unfortunate configEncoding in the bridge. PreCheckCallback does not *see* the secret bits as they get dropped; in https://github.com/pulumi/pulumi-terraform-bridge/blob/e58fb185112b9044616b33801d8cb0194deaf104/pkg/tfbridge/provider.go#L525 when the bridge sets p.configValues they have lost the secret bits and this information simply yes not available here.

**Comment by t0yv0** [comment]
I need to look at this with fresh eyes another day.

**Comment by t0yv0** [comment]
Perhaps state upgrade hooks can have a place to edit state back to correct it before passing down to TF. It does not look clean but can work.

**Comment by t0yv0** [comment]
Noting also that a secrets heuristic that was value and not path-based would work it'd recognize that secret material got copied from "tags" to "tagsAll" without having to code it. But that has some other interesting tradeoffs.

**Comment by t0yv0** [comment]
@iwahbe didn't appreciate merging my PR here. I was working on editing the description and commit message. No harm done for now but in the future let's stick to the engineer initiating the PR merging it.

**Comment by iwahbe** [comment]
@t0yv0 I'm really confused here. I *haven't* been online for the past several hours. I *didn't* merge your PR, though I see that GH says I did.

**Comment by t0yv0** [comment]
That's even more worrisome :)

**Pull request merged**